irohad/torii/command_client.cpp

```diff
@@ -85,47 +67,74 @@ namespace torii {
     return response;
   }
 
+  /**
+   * manages state of an async call.
+   */
+  struct ToriiAsyncClientCall {
+    iroha::protocol::ToriiResponse response;
+    grpc::ClientContext context;
+    grpc::Status status;
+    std::unique_ptr<grpc::ClientAsyncResponseReader<iroha::protocol::ToriiResponse>> response_reader;
+    CommandAsyncClient::Callback callback;
+  };
 
-  /*
-   * TODO(motxx): We can't use CommandClient::ToriiNonBlocking() for now. gRPC causes the error
-   * E0714 04:24:40.045388600    4346 sync_posix.c:60]            assertion failed: pthread_mutex_lock(mu) == 0
+  /**
+   * requests tx to a torii server and returns response (non-blocking)
+   * @param tx
+   * @param callback
    */
-  /*
-  void CommandClient::ToriiNonBlocking(
+  void CommandAsyncClient::Torii(
     const Transaction& tx,
     const std::function<void(ToriiResponse& response)>& callback)
   {
-    ToriiAsyncClientCall* call = new ToriiAsyncClientCall;
+    auto call = new ToriiAsyncClientCall;
+    call->callback = callback;
     call->response_reader = stub_->AsyncTorii(&call->context, tx, &cq_);
     call->response_reader->Finish(&call->response, &call->status, (void*)call);
-
-    listenerPool_->pool.post(std::bind(ToriiNonBlockingListener, cq_, callback));
   }
-  */
 
-  void CommandClient::ToriiNonBlockingListener(
-    grpc::CompletionQueue& cq,
-    const std::function<void(ToriiResponse& response)>& callback)
+  /**
+   * sets ip and port and calls listenToriiNonBlocking() in a new thread.
+   * @param ip
+   * @param port
+   */
+  CommandAsyncClient::CommandAsyncClient(const std::string& ip, const int port)
+    : stub_(iroha::protocol::CommandService::NewStub(
+    grpc::CreateChannel(ip + ":" + std::to_string(port), grpc::InsecureChannelCredentials())))
   {
+    listener_ = std::thread(&CommandAsyncClient::listen, this);
+  }
+
+  CommandAsyncClient::~CommandAsyncClient() {
+    cq_.Shutdown();
+    listener_.join();
+  }
+
+  /**
+   * starts response listener of a non-blocking torii client.
+   */
+  void CommandAsyncClient::listen() {
     void* got_tag;
     bool ok = false;
 
-    while (cq.Next(&got_tag, &ok)) {
-      ToriiAsyncClientCall* call = static_cast<ToriiAsyncClientCall*>(got_tag);
-      assert(ok); // guarantees the request for updates by Finish()
+    while (cq_.Next(&got_tag, &ok)) {
```

Comment:
Maybe add some comments here. It may not be obvious to new people looking at the code what is going on.